hdfs-sink/src/main/java/org/springframework/cloud/stream/module/hdfs/sink/DataStoreWriterFactoryBean.java

```diff
@@ -0,0 +1,222 @@
+ * {@link PartitionTextFileWriter} or {@link TextFileWriter} depending whether partition
+ * path expression is set.
+ *
+ * @author Thomas Risberg
+ * @author Janne Valkealahti
+ * @author Gary Russell
+ */
+public class DataStoreWriterFactoryBean implements InitializingBean, DisposableBean, FactoryBean<DataStoreWriter<String>>,
+		BeanFactoryAware, Lifecycle {
+
+	private HdfsSinkProperties properties;
+
+	private volatile DataStoreWriter<String> storeWriter;
+
+	private volatile Configuration configuration;
+
+	private volatile BeanFactory beanFactory;
+
+
+	@Override
+	public void destroy() throws Exception {
+		storeWriter = null;
+	}
+
+	@Override
+	public DataStoreWriter<String> getObject() throws Exception {
+		return storeWriter;
+	}
+
+	@Override
+	public Class<?> getObjectType() {
+		return DataStoreWriter.class;
+	}
+
+	@Override
+	public boolean isSingleton() {
+		return true;
+	}
+
+	@Override
+	public void afterPropertiesSet() throws Exception {
+		Configuration configurationToUse = null;
+		if (StringUtils.hasText(properties.getFsUri())) {
+			configurationToUse = new Configuration(configuration);
+			configurationToUse.set(CommonConfigurationKeysPublic.FS_DEFAULT_NAME_KEY, properties.getFsUri());
+		}
+		else {
+			configurationToUse = configuration;
+		}
+		EvaluationContext evaluationContext = IntegrationContextUtils.getEvaluationContext(beanFactory);
+		CodecInfo codec = null;
+		if (properties.getCodec() != null) {
+			codec = Codecs.getCodecInfo(properties.getCodec());
+		}
+		List<FileNamingStrategy> strategies = new ArrayList<>();
+		strategies.add(new StaticFileNamingStrategy(properties.getFileName()));
+		if (properties.isFileUuid()) {
+			strategies.add(new UuidFileNamingStrategy());
+		}
+		strategies.add(new RollingFileNamingStrategy());
+		strategies.add(new StaticFileNamingStrategy(properties.getFileExtension(), "."));
+		if (codec != null) {
+			strategies.add(new CodecFileNamingStrategy());
+		}
+		ChainedFileNamingStrategy fileNamingStrategy = new ChainedFileNamingStrategy();
+		fileNamingStrategy.setStrategies(strategies);
+		RolloverStrategy rolloverStrategy = new SizeRolloverStrategy(properties.getRollover());
+		String partitionExpression = null;
+		if (StringUtils.hasText(properties.getPartitionPath())) {
+			partitionExpression = properties.getPartitionPath();
+		}
+		Path basePath = new Path(properties.getDirectory());
+		if (StringUtils.isEmpty(partitionExpression)) {
+			TextFileWriter writer = new TextFileWriter(configurationToUse, basePath, codec);
+			writer.setIdleTimeout(properties.getIdleTimeout());
+			writer.setCloseTimeout(properties.getCloseTimeout());
+			if (StringUtils.hasText(properties.getInUsePrefix())) {
+				((TextFileWriter)storeWriter).setInWritingPrefix(properties.getInUsePrefix());
+			}
+			if (StringUtils.hasText(properties.getInUseSuffix())) {
+				((TextFileWriter)storeWriter).setInWritingSuffix(properties.getInUseSuffix());
```

Comment:
Access to `storeWriter` at this point is always null because it's actually set later in line 142. There seem to be 4 lines which would throw same NPE.